File: frontend/src/types/person.ts
export interface Address {
  line1?: string
  line2?: string
  city?: string
  state?: string
  zip?: string
}

export interface PersonRelationship {
  person_id: number
  related_person_id: number
  relationship_type: string
  created_at: string
  updated_at: string
}

export interface Person {
  id: number
  name: string
  phone_number: string | null
  date_of_birth: string | null
  email: string | null
  employer: string | null
  occupation: string | null
  mailing_address: Address | null
  physical_address: Address | null
  trustor_is_living: number | null
  date_of_death: string | null
  trustor_death_certificate_received: number | null
  trustor_of_sound_mind: number | null
  trustor_has_relinquished: number | null
  trustor_relinquished_date: string | null
  trustor_reling_doc_received: number | null
  created_at: string
  updated_at: string
  is_active: boolean
  has_ssn: boolean
}

export interface PersonWithRelationships extends Person {
  relationships: PersonRelationship[]
}

export interface PersonCreate {
  name: string
  phone_number?: string
  date_of_birth?: string
  ssn?: string
  email?: string
  employer?: string
  occupation?: string
  mailing_address?: Address
  physical_address?: Address
  trustor_is_living?: number
  date_of_death?: string
  trustor_death_certificate_received?: number
  trustor_of_sound_mind?: number
  trustor_has_relinquished?: number
  trustor_relinquished_date?: string
  trustor_reling_doc_received?: number
}

export interface PersonUpdate {
  name?: string
  phone_number?: string
  date_of_birth?: string
  ssn?: string
  email?: string
  employer?: string
  occupation?: string
  mailing_address?: Address
  physical_address?: Address
  is_active?: boolean
  trustor_is_living?: number
  date_of_death?: string
  trustor_death_certificate_received?: number
  trustor_of_sound_mind?: number
  trustor_has_relinquished?: number
  trustor_relinquished_date?: string
  trustor_reling_doc_received?: number
}

export interface PersonRelationshipCreate {
  related_person_id: number
  relationship_type: string
}

export interface PersonListResponse {
  people: Person[]
  total: number
  page: number
  page_size: number
}

File: backend/src/schemas/person.py
from pydantic import BaseModel, EmailStr, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime


class Address(BaseModel):
    """Schema for address as JSON object"""
    line1: Optional[str] = None
    line2: Optional[str] = None
    city: Optional[str] = None
    state: Optional[str] = None
    zip: Optional[str] = None


class PersonRelationshipBase(BaseModel):
    """Base schema for person relationships"""
    related_person_id: int
    relationship_type: str = Field(..., max_length=100, description="Type of relationship (e.g., spouse, child, parent, sibling)")


class PersonRelationshipCreate(PersonRelationshipBase):
    """Schema for creating a person relationship"""
    pass


class PersonRelationship(PersonRelationshipBase):
    """Schema for person relationship response"""
    person_id: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PersonBase(BaseModel):
    """Base schema for Person"""
    name: str = Field(..., min_length=1, max_length=255)
    phone_number: Optional[str] = Field(None, max_length=20)
    date_of_birth: Optional[date] = None
    email: Optional[EmailStr] = None
    employer: Optional[str] = Field(None, max_length=255)
    occupation: Optional[str] = Field(None, max_length=255)
    mailing_address: Optional[Address] = None
    physical_address: Optional[Address] = None
    # Trustor-related fields
    trustor_is_living: Optional[int] = Field(default=1, description="1 = living, 0 = deceased")
    date_of_death: Optional[date] = None
    trustor_death_certificate_received: Optional[int] = Field(default=0)
    trustor_of_sound_mind: Optional[int] = Field(default=1)
    trustor_has_relinquished: Optional[int] = Field(default=0)
    trustor_relinquished_date: Optional[date] = None
    trustor_reling_doc_received: Optional[int] = Field(default=0)


class PersonCreate(PersonBase):
    """Schema for creating a new person"""
    ssn: Optional[str] = Field(None, max_length=11, description="Social Security Number (will be encrypted)")


class PersonUpdate(BaseModel):
    """Schema for updating a person"""
    name: Optional[str] = Field(None, min_length=1, max_length=255)
    phone_number: Optional[str] = Field(None, max_length=20)
    date_of_birth: Optional[date] = None
    email: Optional[EmailStr] = None
    employer: Optional[str] = Field(None, max_length=255)
    occupation: Optional[str] = Field(None, max_length=255)
    mailing_address: Optional[Address] = None
    physical_address: Optional[Address] = None
    ssn: Optional[str] = Field(None, max_length=11, description="Social Security Number (will be encrypted)")
    is_active: Optional[bool] = None
    # Trustor-related fields
    trustor_is_living: Optional[int] = None
    date_of_death: Optional[date] = None
    trustor_death_certificate_received: Optional[int] = None
    trustor_of_sound_mind: Optional[int] = None
    trustor_has_relinquished: Optional[int] = None
    trustor_relinquished_date: Optional[date] = None
    trustor_reling_doc_received: Optional[int] = None


class Person(PersonBase):
    """Schema for person response"""
    id: int
    created_at: datetime
    updated_at: datetime
    is_active: bool
    has_ssn: bool = Field(default=False, description="Indicates if SSN is stored (encrypted)")

    class Config:
        from_attributes = True


class PersonWithRelationships(Person):
    """Schema for person response with relationships"""
    relationships: List[PersonRelationship] = []

    class Config:
        from_attributes = True


class PersonListResponse(BaseModel):
    """Schema for paginated person list response"""
    people: List[Person]
    total: int
    page: int
    page_size: int
